src/object/srv_obj.c

```diff
@@ -2758,9 +2758,55 @@ ds_obj_tgt_punch_handler(crt_rpc_t *rpc)
 	obj_ioc_end(&ioc, rc);
 }
 
+static int
+obj_punch_agg_cb(struct dtx_leader_handle *dlh, void *agg_arg)
+{
+	uint64_t	*flag = agg_arg;
+	int		succeeds = 0;
+	int		allow_failure = 0;
+	int		allow_failure_cnt = 0;
+	int		result = 0;
+	int		i;
+
+	D_ASSERT(flag != NULL);
+	if (*flag & DAOS_COND_PUNCH)
+		allow_failure = -DER_NONEXIST;
+
+	for (i = 0; i < dlh->dlh_sub_cnt; i++) {
+		struct dtx_sub_status	*sub = &dlh->dlh_subs[i];
+
+		if (sub->dss_result == 0) {
+			succeeds++;
+		} else if (sub->dss_result == allow_failure) {
+			allow_failure_cnt++;
+		} else {
+			/* Ignore INPROGRESS if there other failures */
+			if (result == -DER_INPROGRESS || result == 0)
+				result = sub->dss_result;
+		}
+	}
+
+	D_DEBUG(DB_IO, DF_DTI" %d/%d shards flags "DF_X64" result %d\n",
+		DP_DTI(&dlh->dlh_handle.dth_xid), allow_failure_cnt,
+		succeeds, *flag, result);
+
+	if (*flag & DAOS_COND_PUNCH) {
+		/* For punch, if there is one shard succeeds, then
+		 * let's ignore DER_NONEXIST, since the object may not
+		 * exist on some shards due to EC partial udpate.
+		 */
+		if (result == 0 && succeeds == 0)
+			return -DER_NONEXIST;
```

Comment:
should we assert something like allow_failure_cnt == dlh->dlh_sub_cnt ?

Reply:
I can add, but the dlh_sub iteration are in the same function, just a few lines above. :)